### PR TITLE
tidb: always rebuild plan for retry.

### PR DESCRIPTION
**Pull request by coocood**
Fixs record and index inconsistency bug.

### CODE REVIEW

File: ast/misc.go [hunk 1]
```diff
@@ -168,6 +168,7 @@ type ExecuteStmt struct {
 
 	Name      string
 	UsingVars []ExprNode
+	ExecID    uint32
```
Comment (shenli) [inline ast/misc.go]: When will this ExecID be used?

Comment (coocood) [reply ast/misc.go]: To build `Execute` plan.

File: executor/adapter.go [hunk 2]
```diff
@@ -110,7 +110,9 @@ type ExecStmt struct {
 	// Text represents the origin query text.
 	Text string
 
-	ctx            context.Context
+	StmtNode ast.StmtNode
+
+	Ctx            context.Context
```
Comment (zimulala) [inline executor/adapter.go]: Why export it?

Comment (coocood) [reply executor/adapter.go]: It will be constructed by another package.

**Comment by shenli** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**